src/librustdoc/html/highlight.rs

```diff
@@ -727,27 +766,76 @@ fn string<T: Display>(
                 match href {
                     LinkFromSrc::Local(span) => context
                         .href_from_span(*span, true)
-                        .map(|s| format!("{}{}", context_info.root_path, s)),
+                        .map(|s| format!("{}{}", href_context.root_path, s)),
                     LinkFromSrc::External(def_id) => {
-                        format::href_with_root_path(*def_id, context, Some(context_info.root_path))
-                            .ok()
+                        format::href_with_root_path(*def_id, context, Some(href_context.root_path))
                             .map(|(url, _, _)| url)
+                            .or_else(|e| {
+                                if e == format::HrefError::NotInExternalCache
+                                    && matches!(klass, Class::Macro(_))
+                                {
+                                    Ok(generate_macro_def_id_path(href_context, *def_id))
+                                } else {
```

Comment:
Why not incorporate this clause into `format::href_with_root_path` itself? It would require refactoring `href_with_root_path` to take an `HrefContext`, and that would require moving `HrefContext` into `format`, but I think those are both reasonable steps. It would be nice to have a more complete and accurate `href_with_root_path` rather than have a special case that only applies when linkifying.

Reply:
`href_with_root_path` handles external items but the macros are not in the cache used in this function. Hence why I needed to add this function.

Reply:
So what I'm saying is:

 - `href_with_root_path` handles external items in general (functions, structs, etc).
 - `href_with_root_path` does not currently handle external macros.
 - Use cases outside of highlighting don't call `href_with_root_path` for external macros, but that's okay.
 - If `href_with_root_path` could handle external macros just as well as it handles functions and structs, it would be a more complete function.
 - That makes it easier to understand the code base in general.

Reply:
That would force me to pass `Option<HrefContext>` as last argument because the function is called by `href` and `href_with_root_path` is already quite big. Are you sure you don't prefer me creating a wrapper function specific to `highlight` to not make `href_with_root_path` bigger with an extra argument?

Reply:
Actually I don't think you need to add `Option<HrefContext>` as a last argument. Here's the definition of `href_with_root_path`:

```
pub(crate) fn href_with_root_path(
    did: DefId,
    cx: &Context<'_>,
    root_path: Option<&str>,
) -> Result<(String, ItemType, Vec<Symbol>), HrefError> {
```

Here's the definition of `HrefContext`:

```
pub(crate) struct HrefContext<'a, 'b, 'c> {
    pub(crate) context: &'a Context<'b>,
    pub(crate) file_span: Span,
    pub(crate) root_path: &'c str,
}
```

And the code path we're talking about doesn't care about `file_span`, so all it needs are `context` and `root_path`, which are available in `href_with_root_path`.

Reply:
I completely missed that... Sorry for being so slow to catch up and thanks a lot for taking the time to explain to me what was so obvious! Updating it then. Thanks so much for this!

Reply:
It was not so obvious. :-D It took me a couple rounds of reading and thinking "maybe it could work like this..." For instance my original recommendation was to hoist HrefContext to a different module, but that turned out to not be the right path.

Reply:
I made `href_with_root_path` call the function which I moved just above it. I think the end result looks quite nice. It was a great suggestion, thanks a lot for it!